engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java

```diff
@@ -30,22 +29,16 @@ public final class RuntimeCache implements java.util.function.Function<String, O
    */
   @CompilerDirectives.TruffleBoundary
   public boolean offer(UUID key, Object value) {
+    System.err.println("offer " + key + " with " + value);
     var weight = weights.get(key);
     if (weight != null && weight > 0) {
       var ref = new SoftReference<>(value);
       cache.put(key, ref);
       expressions.put(key, new WeakReference<>(value));
-      valuesToKeys.put(value, key);
       return true;
     } else {
-      var otherId = valuesToKeys.get(value);
-      var otherValue = cache.get(otherId);
-      if (otherValue != null && otherValue.get() == value) {
-        // if the value is already cached for another UUID, cache it
-        // for this expression UUID as well
-        var ref = new WeakReference<>(value);
-        expressions.put(key, ref);
-      }
+      var ref = new WeakReference<>(value);
```

Comment:
The idea here: Let's store values of all expressions (with UUID) in the cache, but only as `WeakReference` - as such they can be GCed when nobody else references them. That however means: we need to _evaluate the visualization_ while the argument values are still on stack of the interpreter. The values of all arguments shall be on stack when the last argument is evaluated and the method is about to be called. At that moment we need to evaluate the visualization.

Reply:
It seems that you populate the cache only to provide the value to the observer. Why not trigger the observer (supply the value) directly from the instrumentation without storing it? I feel that this approach will add unnecessary GC pressure and it will be even more noticeable on weaker machines

Reply:
> unnecessary GC pressure

How many lines a typical function has? Less than hundred. How many expressions are on a single line? Less then ten!? We can certainly _handle creation of a thousand_ `WeakReference` objects without seeing any effect on GC. Enso interpreter overhead is way, way bigger...

We can perform some experiment, but I bet the effect is negligible.

Reply:
> populate the cache only to provide the value to the observer. 
> Why not trigger the observer directly from the instrumentation without storing it?

Imagine a [drop down annotation](https://github.com/enso-org/enso/issues/9941#issuecomment-2123993878) in the [demo video](https://github.com/enso-org/enso/pull/10065#issuecomment-2129124396). It is accessing the cache twice:
```ruby
@b for_b
as_list a b = [a, b]

for_b s cache=Nothing =
    a = ["a", cache "a"]
    b = ["b", cache "b"]
```

E.g. the cache is populated to provide **values** (not just a single value) to the observer. E.g. all the values have to be _computed and provided_. Moreover (given the current visualization LS API which associates the visualization only with `self` value), there is no way to know what values from the `cache` will be accessed before running it. Hence we keep all the values in the cache.